src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,910 @@
+        for (let i = 0; i < size; i++) {
+          const key = object['$' + (1 + 2 * i)];
+          const value = object['$' + (1 + (2 * i + 1))];
+          ret.set(key, value);
+        }
+        return ret;
+      }
+
+      case 'XRCore': {
+        const ret = Object.create(null);
+        ret.number = object.$0;
+        ret.name = object.$1;
+        return ret;
+      }
+      default:
+    }
+    return object;
+  });
+  return data;
+}
+
+// This function adds a padding ahead of given string to make its length equal to a given width value
+function zeroPad(s: string, width: number) {
+  return new Array(Math.max(width - s.length, 0) + 1).join('0') + s;
+}
+
+function getOrThrow<K, V>(map: Map<K, V>, k: K): V {
+  if (!map.has(k)) {
+    if (typeof k === 'string') {
+      throw new Error(`Expected key ${k}`);
+    }
+  }
+
+  const value = map.get(k);
+
+  if (value) {
+    return value;
+  }
+
+  throw new Error(`Couldn't find a value`);
+}
+
+function getOrInsert<K, V>(map: Map<K, V>, k: K, fallback: (k: K) => V): V | K {
+  if (!map.has(k)) {
+    map.set(k, fallback(k));
+    return k;
+  }
+  return map.get(k);
+}
+
+// This function extracts arrays of backtraceIDs which contains information about backtrace in recursive manner
+async function getIntegerArrays(
+  samples: InstrumentsSample[],
+  core: TraceDirectoryTree
+): Promise<number[][]> {
+  const uniquing = getOrThrow(core.subdirectories, 'uniquing');
+  const arrayUniquer = getOrThrow(uniquing.subdirectories, 'arrayUniquer');
+  const integeruniquerindex = getOrThrow(
+    arrayUniquer.files,
+    'integeruniquer.index'
+  );
+  const integeruniquerdata = getOrThrow(
+    arrayUniquer.files,
+    'integeruniquer.data'
+  );
+
+  // integeruniquer.index is a binary file containing an array of [byte offset, MB offset] pairs
+  // that indicate where array data starts in the .data file
+
+  // integeruniquer.data is a binary file containing an array of arrays of 64 bit integer.
+  // The schema is a 32 byte header followed by a stream of arrays.
+  // Each array consists of a 4 byte size N followed by N 8 byte little endian integers
+
+  // This table contains the memory addresses of stack frames
+
+  const indexreader = new BinReader(
+    await readAsArrayBuffer(integeruniquerindex)
+  );
+  const datareader = new BinReader(await readAsArrayBuffer(integeruniquerdata));
+
+  // Header we don't care about
+  indexreader.seek(32);
+
+  const arrays: number[][] = [];
+
+  while (indexreader.hasMore()) {
+    const byteOffset =
+      indexreader.readUint32() + indexreader.readUint32() * (1024 * 1024);
+
+    if (byteOffset === 0) {
+      // The first entry in the index table seems to just indicate the offset of
+      // the header into the data file
+      continue;
+    }
+
+    datareader.seek(byteOffset);
+
+    let length = datareader.readUint32();
+    const array: number[] = [];
+
+    while (length--) {
+      array.push(datareader.readUint64());
```

Comment:
what happens if the length isn't consistent with the rest of the data (I mean, if we try to read too much data?)

Reply:
@julienw, I'm not sure what to do here.